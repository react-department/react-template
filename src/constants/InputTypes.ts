enum InputTypes {
  Text = 'TEXT',
  Password = 'PASSWORD',
}

export default InputTypes;
