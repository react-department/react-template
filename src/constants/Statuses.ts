enum Statuses {
  idle = 'idle',
  loading = 'loading',
  failed = 'failed',
}

export default Statuses;
