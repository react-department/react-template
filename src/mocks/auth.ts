export default {
  id: 1,
  token: 'token',
};
