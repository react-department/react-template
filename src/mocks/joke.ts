export default {
  value: 'Chuck Norris can binary search unsorted data.',
};
