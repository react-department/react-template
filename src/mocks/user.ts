export default {
  id: 1,
  firstName: 'John',
  lastName: 'Doe',
};

export const userErrorCredentials = {
  username: 'errorUsername',
  password: 'errorPassword',
};

export const userCredentials = {
  username: 'username',
  password: 'password',
};

export const userError = {
  message: 'Invalid credentials',
};
