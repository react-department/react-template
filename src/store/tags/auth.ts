const prefix = 'AUTH/';
const AUTHENTICATED_USER = `${prefix}AUTHENTICATED_USER`;

export default AUTHENTICATED_USER;
