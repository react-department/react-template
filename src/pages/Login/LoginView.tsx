import { useTranslation } from 'react-i18next';

import ErrorMessage from '../../components/ErrorMessage/ErrorMessage';
import InputTypes from '../../constants/InputTypes';
import useErrorMessage from '../../hooks/useErrorMessage';
import TextInput from './components/TextInput';

import type { ReactElement } from 'react';
import type ILoginView from './interfaces/ILoginView';

import styles from './sass/LoginView.module.scss';

/**
 * Login page view
 */
function LoginView({
  onLoginSubmit,
  errors,
  register,
  isLoading,
  isError,
  error,
}: ILoginView): ReactElement {
  const { t } = useTranslation();
  const errorMessage = useErrorMessage(error);
  return (
    <div className={styles.loginForm}>
      <a href={`${process.env.REACT_APP_API}/users`} target="_blank" rel="noreferrer">
        { t('login.useLoginAndPasswordFromThisJson') }
      </a>
      <form onSubmit={onLoginSubmit}>
        <TextInput
          name="username"
          label={t('login.userName')}
          register={register}
          errors={errors.username}
        />
        <TextInput
          name="password"
          label={t('login.password')}
          register={register}
          errors={errors.password}
          type={InputTypes.Password}
        />
        <input type="submit" disabled={isLoading || Object.keys(errors).length > 0} />
      </form>
      {isError ? <ErrorMessage message={errorMessage} /> : null}
    </div>
  );
}

export default LoginView;
