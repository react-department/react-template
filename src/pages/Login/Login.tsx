import { useForm } from 'react-hook-form';
import { Navigate, useNavigate } from 'react-router-dom';
import { yupResolver } from '@hookform/resolvers/yup';

import { useAppSelector } from '../../store/hooks/useApp';
import { useLoginMutation } from '../../store/slices/auth/apis/dummyAuth';
import { selectUserId } from '../../store/slices/auth/selectors';
import { loginSchema } from '../../utils/validators';
import LoginView from './LoginView';

import type { ReactElement } from 'react';
import type ILoginForm from './interfaces/ILoginForm';

function Login(): ReactElement {
  const navigate = useNavigate();
  const id = useAppSelector(selectUserId);
  const [login, { isLoading, isError, error }] = useLoginMutation();
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<ILoginForm>({
    resolver: yupResolver(loginSchema),
  });
  const onLoginSubmit = handleSubmit(async (credentials): Promise<void> => {
    login(credentials)
      .unwrap()
      .catch(() => {
        reset();
        navigate('/');
      });
  });
  return id ? (
    <Navigate to="/" />
  ) : (
    <LoginView
      onLoginSubmit={onLoginSubmit}
      errors={errors}
      register={register}
      isLoading={isLoading}
      isError={isError}
      error={error}
    />
  );
}

export default Login;
