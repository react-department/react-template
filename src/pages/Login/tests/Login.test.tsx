import { fireEvent, screen, waitFor } from '@testing-library/react';

import { initServer } from '../../../../__tests__/mswServer';
import { render } from '../../../../__tests__/test-utils';
import { testIdErrorMessage, testIdNavigate } from '../../../constants/TestId';
import auth from '../../../mocks/auth';
import { userCredentials, userErrorCredentials } from '../../../mocks/user';
import Login from '../Login';

initServer();

describe('<Login />', () => {
  test('should render component', async () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });
  test('should be redirected if the user was logged in', async () => {
    render(<Login />, { preloadedState: { auth } });
    expect(screen.getByTestId(testIdNavigate)).toBeInTheDocument();
  });
  test('should not be redirected if the user was not logged in', async () => {
    render(<Login />);
    expect(screen.queryByTestId(testIdNavigate)).not.toBeInTheDocument();
  });
  test('should be redirected after login and show error if user set invalid credentials', async () => {
    render(<Login />);
    const nameInput = screen.getByLabelText(/name/i);
    const passwordInput = screen.getByLabelText(/password/i);
    const submitButton = screen.getByRole('button');
    fireEvent.change(nameInput, { target: { value: userErrorCredentials.username } });
    fireEvent.change(passwordInput, { target: { value: userErrorCredentials.password } });
    fireEvent.click(submitButton);
    await waitFor(
      async () => {
        expect(await screen.findByTestId(testIdErrorMessage)).toBeInTheDocument();
      },
    );
    fireEvent.change(nameInput, { target: { value: userCredentials.username } });
    fireEvent.change(passwordInput, { target: { value: userCredentials.password } });
    fireEvent.click(submitButton);
    await waitFor(
      async () => {
        expect(await screen.findByTestId(testIdNavigate)).toBeInTheDocument();
      },
    );
  });
});
