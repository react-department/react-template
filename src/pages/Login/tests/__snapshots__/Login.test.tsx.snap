// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render component 1`] = `
<DocumentFragment>
  <div
    class="loginForm"
  >
    <a
      href="undefined/users"
      rel="noreferrer"
      target="_blank"
    >
      Use login and password from this JSON
    </a>
    <form>
      <label
        class="loginLabel "
        for=":r0:-username"
      >
        <span>
          User Name
        </span>
        <input
          id=":r0:-username"
          name="username"
        />
      </label>
      <label
        class="loginLabel "
        for=":r1:-password"
      >
        <span>
          Password
        </span>
        <input
          id=":r1:-password"
          name="password"
          type="PASSWORD"
        />
      </label>
      <input
        type="submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
